### ISSUE TITLE
Develop Tests

### ISSUE THREAD

**Issue by lloy0076** [Laravel-Backpack/CRUD#676]
I'm thinking tests that:

1. Explore the underlying functionality (i.e. not UX/UI)
2. Explore the UI itself

By underlying functionality:

1. Does every function seem to do what we expect?
2. When storing, does it store?
3. When updating, does it update?

And various combinations (e.g. if storing but the method to not allow storing has been called, do we get what we expect?).

Then for UI:

1. When doing add(Field/Colum) and then rendering, do we get what we expect?
2. Use something to automatically prod the UI (I think Dusk might be the current thing Laravel's using although upstream does tend to keep changing tools).

Because this happens to benefit me, I'm wondering if @tabacitu (or the owner of this package) would mind if I took this on, put together a somewhat more sensible test plan [the above is more a mind dump] and then worked with me to implement this (where I do most of the work, but @tabcitu/someone else makes sure I'm on track/checks things out for me etc)?

**Comment by tabacitu** [comment]
Hi @lloy0076 ,

I would LOVE if you got the ball rolling on this one. I have been focused on getting the product to a complete feature set, always saying I'll write the tests when everything's done. But, as you've seen from the github issues section - new features keep piling up. While that's a very _very_ good thing - the community has been supporting enough to suggest or create some great great features - it does have the downside of having postponed the test again and again.

I think it's best to focus on having full coverage on the CrudPanel methods, and only then test the UI itself. Don't you agree?

Cheers!

**Comment by lloy0076** [comment]
@tabacitu - aye - I was going to suggest the `CrudPanel` methods first for two reasons:

1. If they have bugs/fail then the UI will, at best, be showing "garbage in/garbage out" and at worst a "bug upon a bug"; and
2. Testing the `CrudPanel` methods is probably going to be easier than testing the UI although "Dusk" might make things slightly less painful.

Or to put it another way, if the `CrudPanel` methods are returning crud [pun intended], then even if the UI is perfect it'll be displaying odd stuff :)

**Comment by tabacitu** [comment]
True :-)

I've just sent you an email which should make things easier for you, on the address on your Github profile. Please check it out.

**Comment by tabacitu** [comment]
Totally agree. ```CrudPanel``` methods first. Everything else second. If you get started on this, let me know how I can help.

Cheers!

**Comment by OliverZiegler** [comment]
Anyone started this? Would be nice to verify functionality of progress in #394.
I could also start there writing the tests I need...

**Comment by OliverZiegler** [comment]
So I started writing tests [here](https://github.com/OliverZiegler/CRUD/tree/issue-676).

And already found some inconveniences (see the TODOs).
Would be nice if someone could review this and give some feedback before I proceed.

**Comment by tabacitu** [comment]
@OliverZiegler seems like a good start to me. Looks like you already found things that aren't working properly, which is great :-) 

@tumf87 also said he's interested in this. Maybe a collaboration would be possible?

**Comment by OliverZiegler** [comment]
@tabacitu sounds nice :) 
Should we also fix the issues we find in that branch (so the fixes are connected to the tests) or create separate PRs without the tests? We could also create PRs for single tests or single test classes (like all tests to the columns, to the fields...)

I'm totally open for collaboration @tumf87. Just tell me how you would like to collaborate ;)

**Comment by tabacitu** [comment]
I vote separate PRs - would definitely get merged faster.

**Comment by OliverZiegler** [comment]
Ok then. Will wait for write access as Collaborator and proceed writing tests and fixing code 😁

**Comment by tabacitu** [comment]
Great! Done. 

I've also:
- created the ```unit-tests``` feature branch
- merged your repository into that feature branch
- created a PR that we should NOT merge until it's ready, but the diff might be helpful - https://github.com/Laravel-Backpack/CRUD/pull/828

Anything else I can do to help?

Cheers!

**Comment by OliverZiegler** [comment]
Think thats all for now. Will proceed writing tests then and add PRs with fixes according to failing tests. 😉
